### PR TITLE
Revive Substrate Crate

### PR DESCRIPTION
**Pull request by ggwpez**
Closes https://github.com/paritytech/polkadot-sdk/issues/1450

Bringing back the Substrate crate that was forgotten in the monorepo import 😅. 
It is a doc-only crate. Version number is set to `1.0.0` and publishing is enabled (so that we can link to docs.rs).

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3627756

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3627766

**Review by bkchr — COMMENTED** [review]
IMO we should move the crate into its own folder. It shouldn't at the top of the substrate folder.

**Comment by ggwpez** [comment]
> IMO we should move the crate into its own folder. It shouldn't at the top of the substrate folder.

Where else should it go? This crate is documenting the whole Substrate Framework.  
I can put it into a `doc` folder if you want though.

**Comment by gilescope** [comment]
Then cargo check in the substrate dir would just check substrate code. It's good to have options as to how much code you want to compile. (E.g. you might want to get something working happily with the substrate codebase before then dealing with the impacts to the cumulus and polkadot dirs.) Compiling the kitchen sink would get you this but that's not as intuitive.

File: substrate/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+sc-consensus-babe = { path = "client/consensus/babe" }
+sc-consensus-grandpa = { path = "client/consensus/grandpa" }
+sc-consensus-beefy = { path = "client/consensus/beefy" }
+sc-consensus-manual-seal = { path = "client/consensus/manual-seal" }
+sc-consensus-pow = { path = "client/consensus/pow" }
+
+sp-runtime = { path = "primitives/runtime" }
+frame-support = { path = "frame/support" }
+
+node-cli = { path = "bin/node/cli" }
```
Comment (bkchr) [inline substrate/Cargo.toml:30]:
```suggestion
```

We don't need this anymore. We don't have the binary anymore that required it.

**Comment by bkchr** [comment]
> I can put it into a `doc` folder if you want though.

Good idea 👍

**Comment by bkchr** [comment]
> It's good to have options as to how much code you want to compile

This option is the `-p` parameter to compile exactly what you need. You can also write yourself a script if you really need to have everything compiled in the substrate folder, which is generally just a bad idea. Test it crate locally and then you can check everything.

**Review by kianenigma — APPROVED** [review]
truthfully I prefer root of the folder to 'doc' as it is more aligned with our naming conventions. Otherwise we'd have to make a crate in 'substrate/doc' named 'substrate', which is a new convention now. 

Either way good though

**Comment by bkchr** [comment]
I would have thought we create a top level folder `doc`? We probably want the same doc process for everything.

**Review by liamaharon — APPROVED** [review]

**Review by juangirini — APPROVED** [review]

File: substrate/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+sc-consensus-aura = { path = "client/consensus/aura" }
+sc-consensus-babe = { path = "client/consensus/babe" }
+sc-consensus-grandpa = { path = "client/consensus/grandpa" }
+sc-consensus-beefy = { path = "client/consensus/beefy" }
+sc-consensus-manual-seal = { path = "client/consensus/manual-seal" }
+sc-consensus-pow = { path = "client/consensus/pow" }
+
+sp-runtime = { path = "primitives/runtime" }
+frame-support = { path = "frame/support" }
+
```
Comment (ggwpez) [inline substrate/Cargo.toml:29]:
```suggestion
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3656794

**Comment by ggwpez** [comment]
> I would have thought we create a top level folder `doc`? We probably want the same doc process for everything.

You mean since otherwise it would be inconsistent with the root level? I guess otherwise it would looks like this:
- `/doc/Cargo.toml` `polkadot-sdk` doc crate
- `/substrate/Cargo.toml` ...
- `/polkadot/Cargo.toml` ...
- `/cumulus/Cargo.toml` ...

Yea it would be slightly inconsistent, but that it only since the root level should not have a `src` folder. I think for the other ones it is fine. WDYT?

**Comment by bkchr** [comment]
I mean in general, if we have doc only crates. Which is fine. Why should the not exist in `doc/substrate` or whatever. I'm also not that much interested where we put it, it just "feels" weird to have there a crate in the `substrate` folder.

**Comment by wentelteefje** [comment]
> I mean in general, if we have doc only crates. Which is fine. Why should the not exist in `doc/substrate` or whatever. I'm also not that much interested where we put it, it just "feels" weird to have there a crate in the `substrate` folder.

My idea is to have one top-level crate, which links to the _doc-only_ crates for `substrate`, `polkadot` and `cumulus` (almost like the README of the git repo itself). I think it doesn't really matter where exactly these four crates live in the repository, but I'm not really happy with introducing a `doc` folder at the root level, because then we have both a `docs` and a `doc` (which has some MD-files for contributions etc) folder there. Maybe we can just create four crate folders in `/docs` like so:

`docs/top-level-doc`
`docs/substrate`
`docs/polkadot`
`docs/substrate`

Would that cause any other issues? The user journey I've had in my looks something like this:

`paritytech.github.io/polkadot-sdk` -> `top-level crate` -> User can select between doc-only crates `Substrate`/`Polkadot`/`Cumulus` as their entry point to the projects

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3677751

File: substrate/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+[package]
```
Comment (kianenigma) [inline substrate/Cargo.toml:1]: should be removed.

**Review by kianenigma — COMMENTED** [review]
I am mildly against the folder structure here and want to bikeshed in favor of moving this back to `./substrate/src` in its current structure. 

I just think it is inconsistent with the previous pattern. Also, when we want to create automatic redirects for dead URLs like `https://paritytech.github.io/substrate/master/substrate/index.html`, we would have to make an exception for this. 

Also the naming of what goes in `./doc` and what goes in `./docs` is not clear. 

But happy to move forward as is. 

Now question for @bkchr: if I am to create a new crate that is analogous to the `substrate` crate for `polkadot-sdk`, `polkadot` and `cumulus`, all of thse would also be placed under `./docs`?

**Comment by sam0x17** [comment]
Yeah, I've said this before but principle of least surprise in a rust workspace is that any directory that has crates in it should itself be a crate. This allows for granular compiling based on the folder structure and I would consider it a best practice.

Even when things aren't set up this way, many times people will `cd` into the `substrate` directory, run a cargo command, and have the entirely reasonable expectation that the cargo command should build `substrate`, not _all of `polkadot-sdk`_. It is weird/bad that we don't follow this everywhere, some examples being the `pallets` directory, the `frame` directory, etc etc

**Comment by kianenigma** [comment]
> Yeah, I've said this before but principle of least surprise in a rust workspace, at least for me, would dictate that any directory that has crates in it should itself be a crate containing re-exports of all the child crates (except for private/internal ones). This allows for granular compiling based on the folder structure and I would consider it a best practice.



Yeah, and once we move `./pallets` into a different top level folder (as discussed in Malmo) this ideal that you are posing would be possible: 

```
./frame
   ./src
./pallets
   ./src 
   ./sudo
   ./alliance
   ./...
```

Then I would be happy with having a crate that is just all pallets bundled together. It would be a rather stupid crate to use as you rarely need all pallets, but it would be harmless as it won't pollute the `frame` pallet.

File: Cargo.toml [hunk 4]
```diff
@@ -96,6 +96,7 @@ members = [
 	"cumulus/test/runtime",
 	"cumulus/test/service",
 	"cumulus/xcm/xcm-emulator",
+	"docs/substrate",
```
[the remark this answers is not in the document]
Comment (wentelteefje) [reply Cargo.toml:99]: yes

**Comment by kianenigma** [comment]
> Now question for @bkchr: if I am to create a new crate that is analogous to the substrate crate for polkadot-sdk, polkadot and cumulus, all of thse would also be placed under ./docs?

Also, we want to have a slim version of the master tutorial and reference docs to live in the mono-repo for compatibility and versioning reasons. Where should those live. 

The more I think about this, the more I lean toward: 

1. rename what is called now `docs` to `documents` to signify this is about non-core files (aka _documents_), and avoid confusing with it being interpreted as _documentation_.
2. Move `prdoc` related stuff in there as well cc @chevdor 
3. All umbrella crates such as `polkadot-sdk`, `substrate`, `polkadot` and cumulus should live where they logically belong in file tree.
4. The main tutorial and reference docs should reside under a new folder called `devhub`. So, something like:

```
devhub/
├─ tutorial
│  ├─ step1.rs
│  ├─ step2.rs
│  ├─ see https://github.com/paritytech/polkadot-sdk-docs/issues/3
├─ reference-docs/
│  ├─ extrinsics.rs
│  ├─ account_abstraction.rs
│  ├─ see https://github.com/paritytech/polkadot-sdk-docs/issues/4
documents/
│  ├─ foo.md
│  ├─ bar.adoc
polkadot/
substrate/
cumulus/
```

**Comment by bkchr** [comment]
> All umbrella crates such as polkadot-sdk, substrate, polkadot and cumulus should live where they logically belong in file tree.

If there are only exist for doc reasons, there logical place should be in the docs folder 👀

**Comment by ggwpez** [comment]
I moved it back to `substrate/Cargo.toml` now. Then lets go once the CI is green!

**Comment by kianenigma** [comment]
> moved it back to `substrate/Cargo.toml` now. Then lets go once the CI is green

I am sorry to say that my proposal above was not great and I am about to change my mind 🙈 I think we should move everything into a single crate called `devhub` and let all written documentation be `mod`s in this crate. 

This gives us the ability to navigate backwards and inwards into different crates, and build a nice hierarchy. For example, you could be in `devhub::polkadot_sdk::substrate::frame` and the fact that you can see this path at the top of the page and go backwards is a big win. We cannot really have this nice experience if we have independent crates. 

I think I will explore further with this and possibly change the structure. Once this structure is settled we should finalize the future of markdown files as well #1260.

**Comment by bkchr** [comment]
Lol 😬

**Pull request merged**